trace_replay/io_tracer.cc

```diff
@@ -0,0 +1,239 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "trace_replay/io_tracer.h"
+
+#include <cinttypes>
+#include <cstdio>
+#include <cstdlib>
+
+#include "db/db_impl/db_impl.h"
+#include "db/dbformat.h"
+#include "rocksdb/slice.h"
+#include "util/coding.h"
+#include "util/hash.h"
+#include "util/string_util.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+IOTraceWriter::IOTraceWriter(Env* env, const TraceOptions& trace_options,
+                             std::unique_ptr<TraceWriter>&& trace_writer)
+    : env_(env),
+      trace_options_(trace_options),
+      trace_writer_(std::move(trace_writer)) {}
+
+Status IOTraceWriter::WriteIOOp(const IOTraceRecord& record) {
+  uint64_t trace_file_size = trace_writer_->GetFileSize();
+  if (trace_file_size > trace_options_.max_trace_file_size) {
+    return Status::OK();
+  }
+  Trace trace;
+  trace.ts = record.access_timestamp;
+  trace.type = record.trace_type;
+  Slice file_operation(record.file_operation);
+  PutLengthPrefixedSlice(&trace.payload, file_operation);
+  Slice io_status(record.io_status);
+  PutLengthPrefixedSlice(&trace.payload, io_status);
+  switch (record.trace_type) {
+    case TraceType::kIOGeneral:
+      break;
+    case TraceType::kIOFileSize:
+      PutFixed64(&trace.payload, record.file_size);
+      break;
+    case TraceType::kIOFileNameAndFileSize:
+      PutFixed64(&trace.payload, record.file_size);
+      FALLTHROUGH_INTENDED;
+    case TraceType::kIOFileName: {
+      Slice file_name(record.file_name);
+      PutLengthPrefixedSlice(&trace.payload, file_name);
+      break;
+    }
+    case TraceType::kIOOffset:
+      PutFixed64(&trace.payload, record.offset);
+      break;
+    case TraceType::kIOOffsetAndLen:
+      PutFixed64(&trace.payload, record.offset);
+      FALLTHROUGH_INTENDED;
+    case TraceType::kIOLen:
+      trace.payload.push_back(record.len);
+      break;
+    default:
+      assert(false);
+  }
+  std::string encoded_trace;
+  TracerHelper::EncodeTrace(trace, &encoded_trace);
+  return trace_writer_->Write(encoded_trace);
+}
+
+Status IOTraceWriter::WriteHeader() {
+  Trace trace;
```

Comment:
For the trace header, I suggest to include the following information such that when developer continues change the trace record format, we are able to do backward compatible.
`s << kTraceMagic << "\t"
    << "IO Trace Version: 0.1\t"
    << "RocksDB Version: " << kMajorVersion << "." << kMinorVersion << "\t"
    << "Format: Timestamp OpType Payload\n";`

Reply:
@zhichao-cao , You mean to say I should store header format:  kTraceMagic, kMajorVersion, kMinorVersion in string format instead of binary format?

Reply:
Just ignore my comments here. I thought we use string format for header but I checked the block_cache_tracer, it also uses the binary format. So just keep using binary format.